raiden-ts/src/epics.ts

```diff
@@ -35,6 +54,58 @@ import * as TransportEpics from './transport/epics';
 import * as TransfersEpics from './transfers/epics';
 import * as ServicesEpics from './services/epics';
 
+/**
+ * Returns a shared observable to calculate average block time every [opts.fetchEach] blocks
+ *
+ * @param state$ - Observable of RaidenStates, to fetch blockNumber updates
+ * @param provider - To fetch historic block info
+ * @param opts - Options object
+ * @param opts.initialBlockTime - constant for an initial blockTime (at subscription time)
+ * @param opts.fetchEach - how often to fetch block data (in blocks)
+ * @returns average block time, in milliseconds, from latest [opts.fetchEach] blocks
+ */
+function getBlockTime$(
+  state$: Observable<RaidenState>,
+  provider: RaidenEpicDeps['provider'],
+  { initialBlockTime, fetchEach }: { initialBlockTime: number; fetchEach: number } = {
+    initialBlockTime: 15e3,
+    fetchEach: 10,
+  },
+): Observable<number> {
+  type BlockInfo = readonly [blockNumber: number, timestamp: number, blockTime?: number];
+  let lastInfo: BlockInfo = [-fetchEach, 0]; // previously fetched block info
+
+  // get block info for a given block number
+  const getBlockInfo$ = (blockNumber: number) =>
+    defer(async () =>
+      provider
+        .getBlock(blockNumber)
+        .then((block): BlockInfo => [blockNumber, block.timestamp * 1000]),
+    );
+
+  return concat(
+    // like startWith, but if somehow this gets re-subscribed, uses lastInfo[2] as initialBlockTime
+    defer(() => (lastInfo[0] > 0 && lastInfo[2] ? of(lastInfo[2]) : of(initialBlockTime))),
+    state$.pipe(
+      pluckDistinct('blockNumber'),
+      filter((blockNumber) => blockNumber >= lastInfo[0] + fetchEach),
+      exhaustMap((blockNumber) => {
+        const prevInfo$ =
+          lastInfo[0] > 0 ? of(lastInfo) : getBlockInfo$(Math.max(1, blockNumber - fetchEach));
+        const curInfo$ = getBlockInfo$(blockNumber);
```

Comment:
We do cache it, it'll be `lastInfo` aka `prevInfo$` on next call. We do only a single request every 10 blocks: the current blockInfo, and compare that with the previous request, to calculate it; the only exception for this is  (the triple conditional above), on first request ever, where we need to request both `prevInfo` and `curInfo`, since we haven't any by then

Reply:
Ah, I missed that.